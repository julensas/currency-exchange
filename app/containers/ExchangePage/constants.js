/*
 *
 * ExchangePage constants
 *
 */

export const CLEAR_STATE = 'app/ExchangePage/CLEAR_STATE';
export const FETCH_RATES = 'app/ExchangePage/FETCH_RATES';
export const FETCH_RATES_SUCCESS = 'app/ExchangePage/FETCH_RATES_SUCCESS';
export const FETCH_RATES_FAIL = 'app/ExchangePage/FETCH_RATES_FAIL';
export const LISTEN_TO_RATES = 'app/ExchangePage/LISTEN_TO_RATES';
export const STOP_LISTEN = 'app/ExchangePage/STOP_LISTEN';
export const EXCHANGE = 'app/ExchangePage/EXCHANGE';
