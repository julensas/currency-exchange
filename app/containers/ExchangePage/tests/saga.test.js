/**
 * Tests for HomePage sagas
 */

import { put, take, takeLatest } from 'redux-saga/effects';

import * as C from '../constants';

import exchangeSaga, {
  fetchRates,
  listenToRate,
  listenToRateTask,
} from '../saga';

/* eslint-disable redux-saga/yield-effects */
describe('fetchRates Saga', () => {
  let getInitDataGenerator;

  beforeEach(() => {
    getInitDataGenerator = fetchRates({ payload: 'USD' });

    const selectDescriptor = getInitDataGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getInitDataGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the GET_DATA_SUCCESS action if it requests the data successfully', () => {
    const response = { data: {} };
    const putDescriptor = getInitDataGenerator.next(response).value;
    expect(putDescriptor).toEqual(
      put({
        type: C.FETCH_RATES_SUCCESS,
        payload: {},
      }),
    );
    const putNextDescriptor = getInitDataGenerator.next().value;
    expect(putNextDescriptor).toEqual(
      put({
        type: C.LISTEN_TO_RATES,
        payload: 'USD',
      }),
    );
  });

  it('should call the GET_DATA_FAIL action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getInitDataGenerator.throw(response).value;
    expect(putDescriptor).toEqual(
      put({
        type: C.FETCH_RATES_FAIL,
        payload: response,
      }),
    );
  });
});

describe('listenToRateTask Saga', () => {
  let getInitDataGenerator;

  beforeEach(() => {
    getInitDataGenerator = listenToRateTask({ payload: 'USD' });

    const selectDescriptor = getInitDataGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getInitDataGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the GET_DATA_SUCCESS action if it requests the data successfully', () => {
    const response = { data: {} };
    const putDescriptor = getInitDataGenerator.next(response).value;
    expect(putDescriptor).toEqual(
      put({
        type: C.FETCH_RATES_SUCCESS,
        payload: {},
      }),
    );
  });

  it('should call the GET_DATA_FAIL action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getInitDataGenerator.throw(response).value;
    expect(putDescriptor).toEqual(
      put({
        type: C.FETCH_RATES_FAIL,
        payload: response,
      }),
    );
  });
});

describe('listenToRate Saga', () => {
  let getInitDataGenerator;

  beforeEach(() => {
    getInitDataGenerator = listenToRate({ payload: 'USD' });

    const selectDescriptor = getInitDataGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should take the STOP_LISTEN and cancel listenTask', () => {
    const response = { data: {} };
    const takeDescriptor = getInitDataGenerator.next(response).value;
    expect(takeDescriptor).toEqual(take(C.STOP_LISTEN));
  });
});

describe('products Saga', () => {
  const saga = exchangeSaga();

  it('should start task to watch for FETCH_RATES action', () => {
    const takeLatestDescriptor = saga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(C.FETCH_RATES, fetchRates));
    const takeNextLatestDescriptor = saga.next().value;
    expect(takeNextLatestDescriptor).toEqual(
      takeLatest(C.LISTEN_TO_RATES, listenToRate),
    );
  });
});
