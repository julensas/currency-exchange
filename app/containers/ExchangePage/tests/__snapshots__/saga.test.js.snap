// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchRates Saga should call the GET_DATA_FAIL action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/ExchangePage/STOP_LISTEN",
    },
    "channel": null,
  },
}
`;

exports[`fetchRates Saga should call the GET_DATA_FAIL action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "latest.json",
      Object {
        "params": URLSearchParams {
          Symbol(impl): URLSearchParamsImpl {
            "_list": Array [
              Array [
                "base",
                "USD",
              ],
            ],
            "_url": null,
            Symbol(wrapper): [Circular],
          },
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchRates Saga should dispatch the GET_DATA_SUCCESS action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/ExchangePage/STOP_LISTEN",
    },
    "channel": null,
  },
}
`;

exports[`fetchRates Saga should dispatch the GET_DATA_SUCCESS action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "latest.json",
      Object {
        "params": URLSearchParams {
          Symbol(impl): URLSearchParamsImpl {
            "_list": Array [
              Array [
                "base",
                "USD",
              ],
            ],
            "_url": null,
            Symbol(wrapper): [Circular],
          },
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`listenToRate Saga should take the STOP_LISTEN and cancel listenTask 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "USD",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`listenToRateTask Saga should call the GET_DATA_FAIL action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      10000,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`listenToRateTask Saga should call the GET_DATA_FAIL action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "latest.json",
      Object {
        "params": URLSearchParams {
          Symbol(impl): URLSearchParamsImpl {
            "_list": Array [
              Array [
                "base",
                "[object Object]",
              ],
            ],
            "_url": null,
            Symbol(wrapper): [Circular],
          },
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`listenToRateTask Saga should dispatch the GET_DATA_SUCCESS action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      10000,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`listenToRateTask Saga should dispatch the GET_DATA_SUCCESS action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "latest.json",
      Object {
        "params": URLSearchParams {
          Symbol(impl): URLSearchParamsImpl {
            "_list": Array [
              Array [
                "base",
                "[object Object]",
              ],
            ],
            "_url": null,
            Symbol(wrapper): [Circular],
          },
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
